pennylane/transforms/diagonalize_measurements.py

```diff
@@ -0,0 +1,349 @@
+"""Transform to diagonalize measurements on a tape, assuming all measurements are commuting."""
+
+from copy import copy
+from functools import singledispatch
+
+import pennylane as qml
+from pennylane.operation import Tensor
+from pennylane.ops import CompositeOp, LinearCombination, SymbolicOp
+from pennylane.transforms.core import transform
+
+# pylint: disable=protected-access
+
+_default_supported_obs = frozenset(["PauliZ", "Identity"])
+
+
+def null_postprocessing(results):
+    """A postprocessing function returned by a transform that only converts the batch of results
+    into a result for a single ``QuantumTape``.
+    """
+    return results[0]
+
+
+@transform
+def diagonalize_tape_measurements(tape, supported_base_obs=None):
+    """Diagonalize the measurements on a tape if they are not supported. Raises an error if the
+    measurements do not commute.
+
+    Args:
+        tape (QNode or QuantumScript or Callable): The quantum circuit to modify the measurements of.
+        supported_base_obs (Optional, Iterable(Str)): A list of names of supported base observables.
+            Allowed names are 'PauliX', 'PauliY', 'PauliZ' and 'Hadamard'. If no list is provided,
+            the transform will diagonalize everything.
```

Comment:
Maybe the default should just be `('PauliZ',)`, because by definition we diagonalize to the Z basis?

Reply:
updated to a tuple!